crates/emitter/src/reference_op_emitter.rs

```diff
@@ -0,0 +1,452 @@
+use super::emitter::{AtomIndex, EmitError};
+use crate::ast_emitter::AstEmitter;
+
+#[derive(Debug, PartialEq)]
+enum AssignmentReferenceKind {
+    GlobalName(AtomIndex),
+    #[allow(dead_code)]
+    Prop(AtomIndex),
+    #[allow(dead_code)]
+    Elem,
+}
+
+// See AssignmentReferenceEmitter.
+// This uses struct to hide the details from the consumer.
+#[derive(Debug)]
+#[must_use]
+pub struct AssignmentReference {
+    kind: AssignmentReferenceKind,
+}
+impl AssignmentReference {
+    fn new(kind: AssignmentReferenceKind) -> Self {
+        Self { kind }
+    }
+}
+
+#[derive(Debug, PartialEq)]
+enum CallKind {
+    Normal,
+    // FIXME: Support eval, Function#call, Function#apply etc.
+}
+
+// See *ReferenceEmitter.
+// This uses struct to hide the details from the consumer.
+#[derive(Debug)]
+#[must_use]
+pub struct CallReference {
+    kind: CallKind,
+}
+impl CallReference {
+    fn new(kind: CallKind) -> Self {
+        Self { kind }
+    }
+}
+
+// Struct for emitting bytecode for get `name` operation.
+pub struct GetNameEmitter<'alloc> {
+    pub name: &'alloc str,
+}
+impl<'alloc> GetNameEmitter<'alloc> {
+    pub fn emit(self, emitter: &mut AstEmitter) {
+        let name_index = emitter.emit.get_atom_index(self.name);
+
+        //              [stack]
+
+        // FIXME: Support non-global case.
+        emitter.emit.get_g_name(name_index);
+        //              [stack] VAL
+    }
+}
+
+// Struct for emitting bytecode for get `obj.key` operation.
+pub struct GetPropEmitter<'alloc, F>
+where
+    F: Fn(&mut AstEmitter) -> Result<(), EmitError>,
```

Comment:
Oh, hmm. It's too bad all this requires `AstEmitters`; that means it's tied to the AST.

...It could be done with a trait instead, but that's more complex, and we don't really need it yet. Hmm.

Reply:
I'm about to add trait, as part of scope handling.
that could replace `AstEmitter` reference, with another trait for providing `emit` field.